### Assignment 8
**Name: Amanda Kim**

Problem 1
In this exercise, you will use the Portuguese sea battles data Links to an external site. that contains outcomes of naval battles between Portuguese and Dutch/British ships between 1583 and 1663.
 

The dataset has the following features:

Features:	Features Explained

Battle:	Name of the battle place

Year:	Year of the battle

Portuguese ships:	Number of Portuguese ships

Dutch ships:	Number of Dutch ships

English ships:	Number of ships from English side

The ratio of Portuguese to Dutch/British ships	

Spanish Involvement:	1=Yes, 0=No

Portuguese outcome:	-1=Defeat, 0=Draw, 1=Victory
 

**Instructions**
1. Use an SVM-based model to predict the Portuguese outcome of the battle from the number of ships involved on all sides and Spanish involvement.

2. Try solving the same problem using two other classifiers that you know.

3. Report and compare their results with those from SVM.

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns

df = pd.read_fwf('armada.dat',
                 names=('Battle','Year','Portuguese_ships','Dutch_ships','English_ships','ratio','spanish_involved','portugese_outcome'), 
)
df

,Battle,Year,Portuguese_ships,Dutch_ships,English_ships,ratio,spanish_involved,portugese_outcome
0,Bantam,1601,6,3,0,2.000,0,0
1,Malacca Strait,1606,14,11,0,1.273,0,0
2,Ilha das Naus,1606,6,9,0,0.667,0,-1
3,Pulo Butum,1606,7,9,0,0.778,0,1
4,Surrat,1615,6,0,4,1.500,0,0
5,Ilha das Naus,1615,3,5,0,0.600,0,-1
6,Jask,1620,4,0,4,1.000,0,0
7,Hormuz,1622,6,0,5,1.200,0,-1
8,Mogincoal Shoals,1622,4,4,2,0.667,0,-1
9,Hormuz,1625,8,4,4,1.000,0,0


In [3]:
df.isnull().sum()

Battle               0
Year                 0
Portuguese_ships     0
Dutch_ships          0
English_ships        0
ratio                0
spanish_involved     0
portugese_outcome    0
dtype: int64

In [4]:
#SVM

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

#assigning features & target
X = df[['Portuguese_ships', 'Dutch_ships', 'English_ships','spanish_involved']]
y = df['portugese_outcome']

#test train split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

#scaling data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [5]:
#training svm classifier
svm_classifier = SVC(kernel='linear', C=1.0, random_state=42)
svm_classifier.fit(X_train_scaled, y_train)

#predicting
y_pred = svm_classifier.predict(X_test_scaled)

#print results
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[0 2 0]
 [0 5 0]
 [0 2 0]]

Classification Report:
              precision    recall  f1-score   support

          -1       0.00      0.00      0.00         2
           0       0.56      1.00      0.71         5
           1       0.00      0.00      0.00         2

    accuracy                           0.56         9
   macro avg       0.19      0.33      0.24         9
weighted avg       0.31      0.56      0.40         9



/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(ave

In [6]:
#Logisitic regression
from sklearn.linear_model import LogisticRegression

#training using a multinomial parameter
log_model = LogisticRegression(max_iter=1000, multi_class='multinomial', solver='lbfgs')
log_model.fit(X_train_scaled, y_train)

#predict
y_pred_log = log_model.predict(X_test_scaled)

#results
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_log))
print("\nClassification Report:")
print(classification_report(y_test, y_pred_log))

Confusion Matrix:
[[0 2 0]
 [1 4 0]
 [0 2 0]]

Classification Report:
              precision    recall  f1-score   support

          -1       0.00      0.00      0.00         2
           0       0.50      0.80      0.62         5
           1       0.00      0.00      0.00         2

    accuracy                           0.44         9
   macro avg       0.17      0.27      0.21         9
weighted avg       0.28      0.44      0.34         9



/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:1272: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize

In [7]:
#decision tree model
from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier(max_depth=3,random_state=42)

#train and predict
tree_model.fit(X_train, y_train)
y_pred_tree = tree_model.predict(X_test)

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_tree))
print("\nClassification Report:")
print(classification_report(y_test, y_pred_tree))


Confusion Matrix:
[[1 1 0]
 [2 3 0]
 [1 1 0]]

Classification Report:
              precision    recall  f1-score   support

          -1       0.25      0.50      0.33         2
           0       0.60      0.60      0.60         5
           1       0.00      0.00      0.00         2

    accuracy                           0.44         9
   macro avg       0.28      0.37      0.31         9
weighted avg       0.39      0.44      0.41         9



/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/stevegon/SynologyDrive/code/uw/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(ave

In order to predict Portuguese outcomes in battle based on number of ships and Spanish involvement, I used a SVM model, logistic regression model, and decision tree model. The SVM model had the highest accuracy at 56%, with logistic regression and decision tree both getting 44% accuracy. Though none of these accuracy scores are good, I was surprised that the SVM model had the highest score. However, the decision tree ended up with the highest average F1 score, indicating that the decision tree is more capable of handling the nuances in the dataset compared to the others.
Its important to note that this dataset is very small. I did not know if test-train split was appropriate for this size, but I decided to keep it in order to compare the models to each other better. I do think that small size is why the accuracy score is so low - it is likely that the models overfit the data.

references:
https://www.geeksforgeeks.org/machine-learning/classifying-data-using-support-vector-machinessvms-in-python/